src/cargo/util/context/mod.rs

```diff
@@ -412,6 +413,22 @@ impl GlobalContext {
         self.shell.borrow_mut()
     }
 
+    /// Prints the passed in [Message] to stderr
+    pub fn renderer_diagnostic(&self, message: Message<'_>) -> std::io::Result<()> {
```

Comment:
Any reason that this method is on `GlobalContext` instead of `Shell`?

Reply:
None that I can think of, I went ahead and moved it to `Shell`

Reply:
One of the problems with #12875 is that it put diagnostic tracking in `Shell`.  For me, I see `Shell` as lower level than that.  I feel like putting diagnotic rendering in `Shell` could encourage that kind of thinking.

Reply:
I agree diagnostic tracking should not be in `Shell`, but I am not as worried about encouraging that thinking with this change. This change moves to a function where things get rendered and then emitted, similar to: https://github.com/rust-lang/cargo/blob/ba6fb405284df654362275405175f3fac2b79dbf/src/cargo/core/shell.rs#L387-L393

The JSON gets "rendered" by `serde_json::to_string` and then printed to `stdout`.

The fact `Renderer` gets created inside of `emit_diagnostic` is an implementation detail.

Reply:
@epage do you consider it as a blocker? I am fine with either way.

Reply:
Maybe we could put some design principles in the module-level for shell, so at least this discussion won't just disappear in the air.

Reply:
The current naming is low level enough that I'm fine with moving this forward